pkg/upgrade/upgrade_state.go

```diff
@@ -38,3 +44,167 @@ type ClusterUpgradeState struct {
+	m.Log.V(consts.LogLevelInfo).Info("ProcessReadyOrUnknownNodes")
+
+	for _, nodeState := range currentClusterState.NodeStates[nodeStateName] {
+		if GetPodGeneration(nodeState.DriverPod, m.Log) != nodeState.DriverDaemonSet.GetGeneration() {
+			_ = ChangeNodeUpgradeState(nodeState.Node, UpgradeRequired, m.K8sClient, m.Log)
+			return
+		} else if nodeStateName == Unknown {
+			err := ChangeNodeUpgradeState(nodeState.Node, Ready, m.K8sClient, m.Log)
+			if err != nil {
+				return
+			}
+		}
+		m.Log.V(consts.LogLevelInfo).Info("Node in ready state, upgrade not required",
+			"node", nodeState.Node.Name)
+	}
+}
+
+func (m *ClusterUpgradeStateManager) ProcessUpgradeRequiredNodes(currentClusterState *ClusterUpgradeState, limit int) {
+	m.Log.V(consts.LogLevelInfo).Info("ProcessUpgradeRequiredNodes")
+	for _, nodeState := range currentClusterState.NodeStates[UpgradeRequired] {
+		if limit <= 0 {
+			m.Log.V(consts.LogLevelInfo).Info("Limit for new upgrades is exceeded, skipping the iteration")
+			break
+		}
+
+		err := ChangeNodeUpgradeState(nodeState.Node, Drain, m.K8sClient, m.Log)
+		if err == nil {
+			limit--
+			m.Log.V(consts.LogLevelInfo).Info("Node waiting for drain",
+				"node", nodeState.Node.Name)
+		}
+	}
+}
+
+func (m *ClusterUpgradeStateManager) ProcessDrainNodes(
+	currentClusterState *ClusterUpgradeState, drainSpec *v1alpha1.DrainSpec) {
+	m.Log.V(consts.LogLevelInfo).Info("ProcessDrainNodes")
+
+	// TODO Create drain manager to schedule node drains
+}
+
+func (m *ClusterUpgradeStateManager) ProcessPodRestartNodes(currentClusterState *ClusterUpgradeState) {
+	m.Log.V(consts.LogLevelInfo).Info("ProcessPodRestartNodes")
+
+	pods := make([]*v1.Pod, 0, len(currentClusterState.NodeStates[PodRestart]))
+	for _, nodeState := range currentClusterState.NodeStates[PodRestart] {
+		if GetPodGeneration(nodeState.DriverPod, m.Log) != nodeState.DriverDaemonSet.GetGeneration() {
+			//nolint
+			pods = append(pods, nodeState.DriverPod)
+		} else if nodeState.DriverPod.Status.Phase == "Running" {
+			_ = ChangeNodeUpgradeState(nodeState.Node, Ready, m.K8sClient, m.Log)
+		}
+	}
+
+	// Create pod restart manager to handle pod restarts
+}
+
+func (m *ClusterUpgradeStateManager) ProcessUncordonNodes(currentClusterState *ClusterUpgradeState) {
+	m.Log.V(consts.LogLevelInfo).Info("ProcessUncordonNodes")
+
+	for _, nodeState := range append(
+		currentClusterState.NodeStates[PodRestart], currentClusterState.NodeStates[DrainFailed]...) {
+		if GetPodGeneration(nodeState.DriverPod, m.Log) == nodeState.DriverDaemonSet.GetGeneration() &&
+			nodeState.DriverPod.Status.Phase == "Running" {
+			//nolint
+			if nodeState.Node.Spec.Unschedulable {
+				// TODO uncordon
+			}
+			_ = ChangeNodeUpgradeState(nodeState.Node, Ready, m.K8sClient, m.Log)
+		}
+	}
+}
+
+func GetPodGeneration(pod *v1.Pod, log logr.Logger) int64 {
+	generation, err := strconv.ParseInt(pod.Labels[PodTemplateGenerationLabel], 10, 0)
+	if err != nil {
```

Comment:
error here is really unlikely to happen right ?

because if it does i think we will start draining the node

Reply:
Changed the function to also return an error to be on the safe side